### PR TITLE
Fixing Fix the certman operator logic to look up HZ ID by DNSZone.Status.AWS.ZoneID in Hive

### PR DESCRIPTION
**Pull request by anispate**
[OSD-17701](https://issues.redhat.com/browse/OSD-17701) - Fix the certman operator logic to look up HZ ID by DNSZone.Status.AWS.ZoneID in Hive.

### CODE REVIEW

**Comment by iamkirkbater** [comment]
/approve

I'll give others a chance to look before merging.

File: pkg/clients/aws/route53.go [hunk 1]
```diff
@@ -66,7 +67,9 @@ var fedrampHostedZoneID = os.Getenv(fedrampHostedZoneIDVariable)
 
 // awsClient implements the Client interface
 type awsClient struct {
-	client route53iface.Route53API
+	client     route53iface.Route53API
+	kubeClient client.Client
+	namespace  string
```
Comment (fahlmant) [inline pkg/clients/aws/route53.go:72]: These are unrelated to AWS, so should not be placed inside the AWS Client

File: pkg/clients/aws/route53.go [hunk 2]
```diff
@@ -66,7 +67,9 @@ var fedrampHostedZoneID = os.Getenv(fedrampHostedZoneIDVariable)
 
 // awsClient implements the Client interface
 type awsClient struct {
-	client route53iface.Route53API
+	client     route53iface.Route53API
+	kubeClient client.Client
```
Comment (fahlmant) [inline pkg/clients/aws/route53.go:71]: KubeClient and Namespace are unrelated to AWS, so should not be placed inside the AWS Client

**Review by fahlmant — CHANGES_REQUESTED** [review]
I think the best way forward here to keep things untangled is to refactor into multiple functions
- FindZoneIDForChallenge
This function can handle determining which Zone ID is the right one for a cluster. This is where a kubeclient and the right namespace can be passed as arguments to the function and can be used to find the right hive resource to find which zone to use
- AnswerDNSChallenge
This function keeps the logic of creating the right Resource Record Set and creating it in the target zone, but all the logic for determining which zone to use is removed. It takes the zone ID from FindZoneIDForChallenge as an argument.

This way, the packaging of the kube client and secret don't have to be bundled in with the AWS client, and the logic that interacts with AWS doesn't have to be crossed with the logic that interacts with the kubeapi.

File: controllers/certificaterequest/issue_certificate.go [hunk 3]
```diff
@@ -210,3 +218,23 @@ func (r *CertificateRequestReconciler) IssueCertificate(reqLogger logr.Logger, c
...
+		return "", fmt.Errorf("%d dnsZone objects in a specific namespace found, expected 1 dnsZone", len(dnsZones.Items))
+	}
+
+	if *dnsZones.Items[0].Status.AWS.ZoneID != "" {
+		return filepath.Base(*dnsZones.Items[0].Status.AWS.ZoneID), nil
+	} else if *dnsZones.Items[0].Status.GCP.ZoneName != "" {
+		return *dnsZones.Items[0].Status.GCP.ZoneName, nil
+	} else {
+		return "", nil
+	}
```
Comment (fahlmant) [inline controllers/certificaterequest/issue_certificate.go:239]:
```suggestion
	} 
	
	return "", nil

```

Comment (fahlmant) [reply controllers/certificaterequest/issue_certificate.go:239]: Should this return an error if no Zones are found?

**Review by fahlmant — COMMENTED** [review]
The fedramp and non-fedramp calls to `ChangeResourceRecordSets` look very similar. What's the difference between them, and can we combine them into one block?

File: controllers/certificaterequest/issue_certificate.go (hunk 3 above)
Comment (anispate) [reply controllers/certificaterequest/issue_certificate.go:239]: Yes, you could be right that we can return an error but, I wasn't sure how to perform check ZoneID for the Azure so, the else statement was for Azure.

Comment (fahlmant) [reply controllers/certificaterequest/issue_certificate.go:239]: Does certman-operator support Azure? If so, then you'll need to implement the proper check. If not, then you can return the error and ignore Azure checks.

File: controllers/certificaterequest/issue_certificate.go [hunk 4]
```diff
@@ -210,3 +218,23 @@ func (r *CertificateRequestReconciler) IssueCertificate(reqLogger logr.Logger, c
...
+	err = r.Client.List(context.TODO(), &dnsZones, &client.ListOptions{Namespace: namespace})
+	if err != nil {
+		return "", err
+	}
+
+	if len(dnsZones.Items) != 1 {
+		return "", fmt.Errorf("%d dnsZone objects in a specific namespace found, expected 1 dnsZone", len(dnsZones.Items))
+	}
+
+	if *dnsZones.Items[0].Status.AWS.ZoneID != "" {
```
[the remark this answers is not in the document]
Comment (fahlmant) [reply controllers/certificaterequest/issue_certificate.go:233]: there are more than two possible outcomes
1. AWS is there
2. GCP is there
3. Maybe Azure is there? (according to above comment)
4. None of the above (return err)

File: controllers/certificaterequest/issue_certificate.go (hunk 3 above)
Comment (anispate) [reply controllers/certificaterequest/issue_certificate.go:239]: https://github.com/openshift/hive/blob/master/apis/hive/v1/dnszone_types.go#L149-L150 <<<< Seems like AzureDNSZoneStatus struct is empty so, We don't need Azure condition.

File: controllers/certificaterequest/issue_certificate.go [hunk 5]
```diff
@@ -210,3 +218,22 @@ func (r *CertificateRequestReconciler) IssueCertificate(reqLogger logr.Logger, c
...
+		return filepath.Base(*dnsZones.Items[0].Status.AWS.ZoneID), nil
```
Comment (fahlmant) [inline controllers/certificaterequest/issue_certificate.go:234]: Why is `filepath.Base` needed here?

File: controllers/certificaterequest/issue_certificate.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rafael-azevedo) [reply controllers/certificaterequest/issue_certificate.go:233]: @anispate I am not sure you addressed the concerns here. I'd either remove the else statements or use a case switch. The way I am reading this its either AWS or GCP and only one of these can happen

File: pkg/clients/aws/route53.go [hunk 6]
```diff
@@ -73,107 +72,48 @@ func (c *awsClient) GetDNSName() string {
...
+		zone, err := c.client.GetHostedZone(&route53.GetHostedZoneInput{Id: &fedrampHostedZoneID})
 		if err != nil {
-			reqLogger.Error(err, result.GoString(), "fqdn", fqdn)
+			reqLogger.Error(err, err.Error())
 			return "", err
 		}
-
-		return fqdn, nil
+		input.HostedZoneId = zone.HostedZone.Id
+	} else {
```
Comment (rafael-azevedo) [inline pkg/clients/aws/route53.go:107]: instead of the else here you can set input.HostedZoneId = &dnsZone and overwrite it if it is fedramp

**Review by rafael-azevedo — CHANGES_REQUESTED** [review]

File: controllers/certificaterequest/issue_certificate.go [hunk 7]
```diff
@@ -210,3 +218,22 @@ func (r *CertificateRequestReconciler) IssueCertificate(reqLogger logr.Logger, c
...
+	if err != nil {
+		return "", err
+	}
+
+	if len(dnsZones.Items) != 1 {
+		return "", fmt.Errorf("%d dnsZone objects in a specific namespace found, expected 1 dnsZone", len(dnsZones.Items))
+	}
+
+	if *dnsZones.Items[0].Status.AWS.ZoneID != "" {
+		return filepath.Base(*dnsZones.Items[0].Status.AWS.ZoneID), nil //the format of that ID is (/hostedzone/<something>) so, we don't care about >/. Here, we only want to grab the stuff after the last /.
```
Comment (rafael-azevedo) [inline controllers/certificaterequest/issue_certificate.go:234]: The comments above were not yet resolved I'd either remove the else statements or use a case switch. The way I am reading this its either AWS or GCP and only one of these can happen

File: controllers/certificaterequest/issue_certificate.go [hunk 8]
```diff
@@ -210,3 +218,22 @@ func (r *CertificateRequestReconciler) IssueCertificate(reqLogger logr.Logger, c
...
+		return "", err
+	}
+
+	if len(dnsZones.Items) != 1 {
+		return "", fmt.Errorf("%d dnsZone objects in a specific namespace found, expected 1 dnsZone", len(dnsZones.Items))
+	}
+
+	if *dnsZones.Items[0].Status.AWS.ZoneID != "" {
+		return filepath.Base(*dnsZones.Items[0].Status.AWS.ZoneID), nil //the format of that ID is (/hostedzone/<something>) so, we don't care about >/. Here, we only want to grab the stuff after the last
+	} else if *dnsZones.Items[0].Status.GCP.ZoneName != "" {
```
Comment (fahlmant) [inline controllers/certificaterequest/issue_certificate.go:235]:
```suggestion
	}
	if *dnsZones.Items[0].Status.GCP.ZoneName != "" {
```

**Review by fahlmant — APPROVED** [review]
/lgtm

File: controllers/certificaterequest/issue_certificate.go [hunk 9]
```diff
@@ -210,3 +218,23 @@ func (r *CertificateRequestReconciler) IssueCertificate(reqLogger logr.Logger, c
...
+	dnsZones := hivev1.DNSZoneList{}
+	err = r.Client.List(context.TODO(), &dnsZones, &client.ListOptions{Namespace: namespace})
+	if err != nil {
+		return "", err
+	}
+
+	if len(dnsZones.Items) != 1 {
+		return "", fmt.Errorf("%d dnsZone objects in a specific namespace found, expected 1 dnsZone", len(dnsZones.Items))
+	}
+
```
Comment (iamkirkbater) [inline controllers/certificaterequest/issue_certificate.go:232]:
```
dnsZoneStatus := *dnsZones.Items[0].Status

if dnsZoneStatus == nil {
  return "", fmt.Error("Unexpected error, status is not populated")
}

if dnsZoneStatus.AWS != nil {
   // its aws, do something
   return ...
 }
 
 if dnsZoneStatus.GCP != nil {
   // its gcp, do something else
   return ...
 }
 
 return "", fmt.Errorf("Unexpected error: not aws or gcp don't know what to do here")
 ```
